### ISSUE TITLE
protocol statement in id section

### ISSUE THREAD

**Issue by brentzundel** [w3c/vc-data-model#478]
In [section 4.2](https://w3c.github.io/vc-data-model/CR/2019-03-26/#identifiers), there is a protocol statement that incorrectly uses normative language that cannot be enforced.

The text reads:
>  Identifiers that others are expected to use when expressing statements about a specific thing MUST be expressed using the id property.

**This does not match the pattern of all other normative statements in the spec, and therefore I argue that this is actually an editorial change.**

To follow the pattern of all other statements of this type, the text should read
> This specification defines the id property for identifiers. This allows for the expression of statements about specific things in the verifiable credential.

**Comment by David-Chadwick** [comment]
Brent, whilst I don't object to your rewording, why is this a protocol statement? Surely this is a modelling statement. It is saying that if an object has an identifier that others use when referring to this object, then it must be the ID property.

Note that we have two different types of normative statements about properties:
1. Properties that may be present, but if present MUST have a specific format, and
2. Properties that MUST be present, and MUST have a specific format.

The ID property is rather unusual in that both of the above apply to it e.g. a VC MUST have an ID property, whereas a CredentialSubject may have an ID property. 

What is missing from the spec, is a statement that says "A verifiable credential MUST have an ID property". I think the sentence you have identified was a proxy for this missing sentence!

Another related point is that the NOTE implies that a VC can omit the ID property, and I do not think this is intentional. So the NOTE is dangerous as it implies that no IDs are mandatory (even ones that are stated to be).

In conclusion, I think this entire section needs some re-writing and clarification.

**Comment by brentzundel** [comment]
1. According to the spec, an id property is entirely optional.
2. The normative statement here is describing how an id property must be **used**, which is protocol and impossible to test as a data model.

**Comment by ken-ebert** [comment]
As I remember from previous discussions, the intent was that if the optional identifiers were present, then they MUST be contained in the "id" and not in some other attribute.

I think this language might capture that thought better:
"Identifiers that others are expected to use when expressing statements about a specific thing, if present, MUST be expressed using the id property."

I agree with Brent that it might be very difficult to test whether the issuer used another attribute for an identifier instead of "id", such as:
"myid" // relatively easy to spot
or
"my-noncompliant-identity" // relatively easy to spot
or
"qxz" // should this have been "id"?

Although I think this is very difficult to test, I don't think this is a protocol issue. I think it is a question related to the basic filling out of the data model.

**Comment by brentzundel** [comment]
@David-Chadwick and @ken-ebert I will submit a PR that addresses this issue. I will use the language above that you've said you don't object to and not worry about the misunderstandings between us as they don't really stall the ability to resolve this in a way we can all agree on.

**Comment by David-Chadwick** [comment]
Your PR does not correctly address this issue. I will comment further against the PR

**Comment by brentzundel** [comment]
I have modified PR #534, please re-review.

**Comment by msporny** [comment]
PR #534 has been merged.

**Comment by msporny** [comment]
The next step is to see if the WG agrees with the change to the specification text.

**Comment by msporny** [comment]
@brentzundel this language is testable by checking to ensure that when an `id` property is provided, that the processor doesn't modify or remove it. This is tested in the current test suite in this test:

https://github.com/w3c/vc-test-suite/blob/gh-pages/test/vc-data-model-1.0/10-basic.js#L50

**Comment by msporny** [comment]
The next step is to see if the WG agrees with the change to the specification text.

**Comment by TallTed** [comment]
> this language is testable by checking to ensure that when an `id` property is provided, that the processor doesn't remove it. This is tested in the current test suite in [this test](https://github.com/w3c/vc-test-suite/blob/gh-pages/test/vc-data-model-1.0/10-basic.js#L50)

A model has no processors, so there's no process to test.

As I read the tests at that link, they're checking that if `id` is present, it's a single URI.  There's nothing about a processor, nor any removal or retention...

Also, #534 had significant wording issues, to my reading.  Suggested revision is now (belatedly) a comment on #534.

**Comment by brentzundel** [comment]
Closing this issue, as the PR satisfying it has been merged.
